E2E test: Server-initiated close handling

## Problem

`StreamConnection::dispatchServerPush()` handles server-initiated close (key `0x0016`) by sending a CloseResponse and closing the connection. However, **no E2E test verifies this behavior**. This is a critical path — in production, the server may close connections due to policy violations, resource limits, or administrative actions.

## How to trigger

Server-initiated close can be triggered by:
1. Deleting a stream while a consumer is subscribed to it (may trigger MetadataUpdate + Close)
2. Using the RabbitMQ management API to force-close a connection
3. Server shutdown/restart

## Expected test

```php
public function testServerInitiatedCloseIsHandledGracefully(): void
{
    $connection = Connection::create(...);
    $streamName = 'test-server-close-' . uniqid();
    $connection->createStream($streamName);

    // Force close via management API
    // curl -X DELETE http://localhost:15672/api/connections/...

    // Next operation should detect the closed connection
    $this->expectException(\RuntimeException::class);
    $this->expectExceptionMessage('Connection closed');
    $connection->createStream('another-stream');
}
```

## Why it matters

- Critical production scenario — servers close connections regularly
- The `dispatchServerPush()` Close handler sends a response and calls `$this->close()`, but this path is untested
- Without this test, regressions in close handling could cause silent data loss

## Acceptance criteria

- [ ] Verify `isConnected()` returns `false` after server-initiated close
- [ ] Verify subsequent operations throw meaningful exceptions
- [ ] Verify the client sends a proper CloseResponse before disconnecting

## CI Pipeline Failed

Pipeline checks failed. Logs saved to `/home/decodo/work/rabbit-stream/.oda/artifacts/151/pipeline-fail.log`.

```
## Failed CI Checks

Failed: Lint (PHPCS + Rector + PHPStan)

### Run 23748556655

```
Lint (PHPCS + Rector + PHPStan)	Run Rector (dry-run)	﻿2026-03-30T13:57:38.0688013Z ##[group]Run composer rector
Lint (PHPCS + Rector + PHPStan)	Run Rector (dry-run)	2026-03-30T13:57:38.0688323Z [36;1mcomposer rector[0m
Lint (PHPCS + Rector + PHPStan)	Run Rector (dry-run)	2026-03-30T13:57:38.0715147Z shell: /usr/bin/bash -e {0}
Lint (PHPCS + Rector + PHPStan)	Run Rector (dry-run)	2026-03-30T13:57:38.0715388Z env:
Lint (PHPCS + Rector + PHPStan)	Run Rector (dry-run)	2026-03-30T13:57:38.0715567Z   COMPOSER_PROCESS_TIMEOUT: 0
Lint (PHPCS + Rector + PHPStan)	Run Rector (dry-run)	2026-03-30T13:57:38.0715815Z   COMPOSER_NO_INTERACTION: 1
Lint (PHPCS + Rector + PHPStan)	Run Rector (dry-run)	2026-03-30T13:57:38.0716037Z   COMPOSER_NO_AUDIT: 1
Lint (PHPCS + Rector + PHPStan)	Run Rector (dry-run)	2026-03-30T13:57:38.0716243Z ##[endgroup]
Lint (PHPCS + Rector + PHPStan)	Run Rector (dry-run)	2026-03-30T13:57:46.9257403Z    0/213 [░░░░░░░░░░░░░░░░░░░░░░░░░░░░]   0%
Lint (PHPCS + Rector + PHPStan)	Run Rector (dry-run)	2026-03-30T13:57:47.0158283Z 
Lint (PHPCS + Rector + PHPStan)	Run Rector (dry-run)	2026-03-30T13:57:47.0158841Z 1 file with changes
Lint (PHPCS + Rector + PHPStan)	Run Rector (dry-run)	2026-03-30T13:57:47.0159552Z ===================
Lint (PHPCS + Rector + PHPStan)	Run Rector (dry-run)	2026-03-30T13:57:47.0159830Z 
Lint (PHPCS + Rector + PHPStan)	Run Rector (dry-run)	2026-03-30T13:57:47.0161911Z 1) tests/E2E/ServerInitiatedCloseTest.php:152
Lint (PHPCS + Rector + PHPStan)	Run Rector (dry-run)	2026-03-30T13:57:47.0162349Z 
Lint (PHPCS + Rector + PHPStan)	Run Rector (dry-run)	2026-03-30T13:57:47.0182327Z     ---------- begin diff ----------
Lint (PHPCS + Rector + PHPStan)	Run Rector (dry-run)	2026-03-30T13:57:47.0182688Z @@ @@
Lint (PHPCS + Rector + PHPStan)	Run Rector (dry-run)	2026-03-30T13:57:47.0182905Z          try {
Lint (PHPCS + Rector + PHPStan)	Run Rector (dry-run)	2026-03-30T13:57:47.0183223Z              $connection->createStream('stream-1-' . uniqid());
Lint (PHPCS + Rector + PHPStan)	Run Rector (dry-run)	2026-03-30T13:57:47.0183749Z              $this->fail('Expected ConnectionException was not thrown');
Lint (PHPCS + Rector + PHPStan)	Run Rector (dry-run)	2026-03-30T13:57:47.0184300Z -        } catch (ConnectionException $e) {
Lint (PHPCS + Rector + PHPStan)	Run Rector (dry-run)	2026-03-30T13:57:47.0184677Z +        } catch (ConnectionException) {
Lint (PHPCS + Rector + PHPStan)	Run Rector (dry-run)	2026-03-30T13:57:47.0185001Z              // Expected
Lint (PHPCS + Rector + PHPStan)	Run Rector (dry-run)	2026-03-30T13:57:47.0185262Z          }
Lint (PHPCS + Rector + PHPStan)	Run Rector (dry-run)	2026-03-30T13:57:47.0185388Z 
Lint (PHPCS + Rector + PHPStan)	Run Rector (dry-run)	2026-03-30T13:57:47.0185476Z @@ @@
Lint (PHPCS + Rector + PHPStan)	Run Rector (dry-run)	2026-03-30T13:57:47.0185742Z          // Find our connection by peer_address
Lint (PHPCS + Rector + PHPStan)	Run Rector (dry-run)	2026-03-30T13:57:47.0186106Z          $ourConnection = null;
Lint (PHPCS + Rector + PHPStan)	Run Rector (dry-run)	2026-03-30T13:57:47.0186423Z          foreach ($connections as $conn) {
Lint (PHPCS + Rector + PHPStan)	Run Rector (dry-run)	2026-03-30T13:57:47.0186891Z -            if (isset($conn['peer_address']) && $conn['peer_address'] === $localIp) {
Lint (PHPCS + Rector + PHPStan)	Run Rector (dry-run)	2026-03-30T13:57:47.0187475Z -                // Also verify it's a stream connection on the right port
Lint (PHPCS + Rector + PHPStan)	Run Rector (dry-run)	2026-03-30T13:57:47.0188002Z -                if (isset($conn['peer_port']) && $conn['peer_port'] > 0) {
Lint (PHPCS + Rector + PHPStan)	Run Rector (dry-run)	2026-03-30T13:57:47.0188443Z -                    $ourConnection = $conn;
Lint (PHPCS + Rector + PHPStan)	Run Rector (dry-run)	2026-03-30T13:57:47.0188773Z -                    break;
Lint (PHPCS + Rector + PHPStan)	Run Rector (dry-run)	2026-03-30T13:57:47.0189313Z -                }
Lint (PHPCS + Rector + PHPStan)	Run Rector (dry-run)	2026-03-30T13:57:47.0189640Z +            if (!isset($conn['peer_address'])) {
Lint (PHPCS + Rector + PHPStan)	Run Rector (dry-run)	2026-03-30T13:57:47.0189981Z +                continue;
Lint (PHPCS + Rector + PHPStan)	Run Rector (dry-run)	2026-03-30T13:57:47.0190247Z              }
Lint (PHPCS + Rector + PHPStan)	Run Rector (dry-run)	2026-03-30T13:57:47.0190514Z +            if ($conn['peer_address'] !== $localIp) {
Lint (PHPCS + Rector + PHPStan)	Run Rector (dry-run)	2026-03-30T13:57:47.0190875Z +                continue;
Lint (PHPCS + Rector + PHPStan)	Run Rector (dry-run)	2026-03-30T13:57:47.0191163Z +            }
Lint (PHPCS + Rector + PHPStan)	Run Rector (dry-run)	2026-03-30T13:57:47.0191414Z +            if (!isset($conn['peer_port'])) {
Lint (PHPCS + Rector + PHPStan)	Run Rector (dry-run)	2026-03-30T13:57:47.0191753Z +                continue;
Lint (PHPCS + Rector + PHPStan)	Run Rector (dry-run)	2026-03-30T13:57:47.0192005Z +            }
Lint (PHPCS + Rector + PHPStan)	Run Rector (dry-run)	2026-03-30T13:57:47.0192263Z +            if ($conn['peer_port'] <= 0) {
Lint (PHPCS + Rector + PHPStan)	Run Rector (dry-run)	2026-03-30T13:57:47.0192579Z +                continue;
Lint (PHPCS + Rector + PHPStan)	Run Rector (dry-run)	2026-03-30T13:57:47.0192837Z +            }
Lint (PHPCS + Rector + PHPStan)	Run Rector (dry-run)	2026-03-30T13:57:47.0193076Z +            $ourConnection = $conn;
Lint (PHPCS + Rector + PHPStan)	Run Rector (dry-run)	2026-03-30T13:57:47.0193389Z +            break;
Lint (PHPCS + Rector + PHPStan)	Run Rector (dry-run)	2026-03-30T13:57:47.0193622Z          }
Lint (PHPCS + Rector + PHPStan)	Run Rector (dry-run)	2026-03-30T13:57:47.0193753Z 
Lint (PHPCS + Rector + PHPStan)	Run Rector (dry-run)	2026-03-30T13:57:47.0193867Z          if ($ourConnection === null) {
Lint (PHPCS + Rector + PHPStan)	Run Rector (dry-run)	2026-03-30T13:57:47.0194306Z              // Try alternative: close all stream connections from our host
Lint (PHPCS + Rector + PHPStan)	Run Rector (dry-run)	2026-03-30T13:57:47.0194767Z              foreach ($connections as $conn) {
Lint (PHPCS + Rector + PHPStan)	Run Rector (dry-run)	2026-03-30T13:57:47.0195246Z -                if (isset($conn['peer_address']) && $conn['peer_address'] === $localIp) {
Lint (PHPCS + Rector + PHPStan)	Run Rector (dry-run)	2026-03-30T13:57:47.0195834Z -                    if (isset($conn['protocol']) && $conn['protocol'] === 'stream') {
Lint (PHPCS + Rector + PHPStan)	Run Rector (dry-run)	2026-03-30T13:57:47.0196601Z -                        $ourConnection = $conn;
Lint (PHPCS + Rector + PHPStan)	Run Rector (dry-run)	2026-03-30T13:57:47.0196937Z -                        break;
Lint (PHPCS + Rector + PHPStan)	Run Rector (dry-run)	2026-03-30T13:57:47.0197221Z -                    }
Lint (PHPCS + Rector + PHPStan)	Run Rector (dry-run)	2026-03-30T13:57:47.0197511Z +                if (!isset($conn['peer_address'])) {
Lint (PHPCS + Rector + PHPStan)	Run Rector (dry-run)	2026-03-30T13:57:47.0197858Z +                    continue;
Lint (PHPCS + Rector + PHPStan)	Run Rector (dry-run)	2026-03-30T13:57:47.0198134Z                  }
Lint (PHPCS + Rector + PHPStan)	Run Rector (dry-run)	2026-03-30T13:57:47.0198416Z +                if ($conn['peer_address'] !== $localIp) {
Lint (PHPCS + Rector + PHPStan)	Run Rector (dry-run)	2026-03-30T13:57:47.0198780Z +                    continue;
Lint (PHPCS + Rector + PHPStan)	Run Rector (dry-run)	2026-03-30T13:57:47.0199036Z +                }
Lint (PHPCS + Rector + PHPStan)	Run Rector (dry-run)	2026-03-30T13:57:47.0199603Z +                if (!isset($conn['protocol'])) {
Lint (PHPCS + Rector + PHPStan)	Run Rector (dry-run)	2026-03-30T13:57:47.0200354Z +                    continue;
Lint (PHPCS + Rector + PHPStan)	Run Rector (dry-run)	2026-03-30T13:57:47.0200644Z +                }
Lint (PHPCS + Rector + PHPStan)	Run Rector (dry-run)	2026-03-30T13:57:47.0200878Z +                if ($conn['protocol'] !== 'stream') {
Lint (PHPCS + Rector + PHPStan)	Run Rector (dry-run)	2026-03-30T13:57:47.0201171Z +                    continue;
Lint (PHPCS + Rector + PHPStan)	Run Rector (dry-run)	2026-03-30T13:57:47.0201407Z +                }
Lint (PHPCS + Rector + PHPStan)	Run Rector (dry-run)	2026-03-30T13:57:47.0201621Z +                $ourConnection = $conn;
Lint (PHPCS + Rector + PHPStan)	Run Rector (dry-run)	2026-03-30T13:57:47.0201890Z +                break;
Lint (PHPCS + Rector + PHPStan)	Run Rector (dry-run)	2026-03-30T13:57:47.0202086Z              }
Lint (PHPCS + Rector + PHPStan)	Run Rector (dry-run)	2026-03-30T13:57:47.0202267Z          }
Lint (PHPCS + Rector + PHPStan)	Run Rector (dry-run)	2026-03-30T13:57:47.0202464Z     ----------- end diff -----------
Lint (PHPCS + Rector + PHPStan)	Run Rector (dry-run)	2026-03-30T13:57:47.0202644Z 
Lint (PHPCS + Rector + PHPStan)	Run Rector (dry-run)	2026-03-30T13:57:47.0202745Z Applied rules:
Lint (PHPCS + Rector + PHPStan)	Run Rector (dry-run)	2026-03-30T13:57:47.0203713Z  * SimplifyConditionsRector
Lint (PHPCS + Rector + PHPStan)	Run Rector (dry-run)	2026-03-30T13:57:47.0204193Z  * ChangeNestedForeachIfsToEarlyContinueRector
Lint (PHPCS + Rector + PHPStan)	Run Rector (dry-run)	2026-03-30T13:57:47.0204743Z  * RemoveUnusedVariableInCatchRector
Lint (PHPCS + Rector + PHPStan)	Run Rector (dry-run)	2026-03-30T13:57:47.0204959Z 
Lint (PHPCS + Rector + PHPStan)	Run Rector (dry-run)	2026-03-30T13:57:47.0204965Z 
Lint (PHPCS + Rector + PHPStan)	Run Rector (dry-run)	2026-03-30T13:57:47.0221711Z  [OK] 1 file would have been changed (dry-run) by Rector                        
Lint (PHPCS + Rector + PHPStan)	Run Rector (dry-run)	2026-03-30T13:57:47.0222212Z 
Lint (PHPCS + Rector + PHPStan)	Run Rector (dry-run)	2026-03-30T13:57:47.0334688Z  213/213 [▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓] 100%Script rector process --dry-run handling the rector event returned with error code 2
Lint (PHPCS + Rector + PHPStan)	Run Rector (dry-run)	2026-03-30T13:57:47.0436134Z ##[error]Process completed with exit code 2.
```


```

--- a/src/StreamConnection.php
+++ b/src/StreamConnection.php
@@ -524,12 +524,17 @@ class StreamConnection
 
         while ($remaining > 0) {
             $chunk = socket_read($this->socket, $remaining);
-            if ($chunk === false || $chunk === '') {
+            if ($chunk === false) {
                 $error = socket_last_error($this->socket);
                 if ($error === SOCKET_ETIMEDOUT) {
                     return null;
                 }
+                $this->connected = false;
                 throw new ConnectionException("Failed to read from socket: " . socket_strerror($error));
+            }
+            if ($chunk === '') {
+                $this->connected = false;
+                throw new ConnectionException("Failed to read from socket: connection closed by peer");
             }
 
             $data .= $chunk;

--- a/tests/E2E/ServerInitiatedCloseTest.php
+++ b/tests/E2E/ServerInitiatedCloseTest.php
@@ -83,7 +83,10 @@ class ServerInitiatedCloseTest extends TestCase
             }
         }
 
-        $this->assertNotNull($lastException, 'Expected ConnectionException was not thrown after server-initiated close');
+        $this->assertNotNull(
+            $lastException,
+            'Expected ConnectionException was not thrown after server-initiated close'
+        );
 
         // After the operation fails, connection should be marked as disconnected
         $this->assertFalse($connection->isConnected());

--- a/tests/E2E/ServerInitiatedCloseTest.php
+++ b/tests/E2E/ServerInitiatedCloseTest.php
@@ -1,0 +1,184 @@
+<?php
+
+declare(strict_types=1);
+
+namespace CrazyGoat\RabbitStream\Tests\E2E;
+
+use CrazyGoat\RabbitStream\Client\Connection;
+use CrazyGoat\RabbitStream\Exception\ConnectionException;
+use PHPUnit\Framework\TestCase;
+
+/**
+ * @group destructive
+ */
+class ServerInitiatedCloseTest extends TestCase
+{
+    private static string $host = '127.0.0.1';
+    private static int $port = 5552;
+    private static int $managementPort = 15672;
+
+    private ?Connection $connection = null;
+    private string $streamName;
+
+    public static function setUpBeforeClass(): void
+    {
+        self::$host = getenv('RABBITMQ_HOST') ?: self::$host;
+        self::$port = (int)(getenv('RABBITMQ_PORT') ?: self::$port);
+        self::$managementPort = (int)(getenv('RABBITMQ_MANAGEMENT_PORT') ?: self::$managementPort);
+    }
+
+    protected function setUp(): void
+    {
+        $this->connection = Connection::create(self::$host, self::$port);
+        $this->streamName = 'test-srv-close-' . uniqid();
+        $this->connection->createStream($this->streamName);
+    }
+
+    protected function tearDown(): void
+    {
+        // Try to clean up the stream via a new connection if our connection was closed
+        try {
+            $cleanupConn = Connection::create(self::$host, self::$port);
+            $cleanupConn->deleteStream($this->streamName);
+            $cleanupConn->close();
+        } catch (\Exception) {
+            // Ignore cleanup errors
+        }
+
+        if ($this->connection instanceof Connection) {
+            try {
+                $this->connection->close();
+            } catch (\Exception) {
+                // Ignore cleanup errors - connection may already be closed
+            }
+        }
+    }
+
+    public function testServerInitiatedCloseIsHandledGracefully(): void
+    {
+        $connection = $this->connection;
+        $this->assertNotNull($connection);
+
+        // Find our connection name in RabbitMQ management API
+        // The management API may need a few seconds to register the connection
+        $connectionName = $this->getStreamConnectionName();
+        $this->assertNotNull($connectionName, 'Could not find stream connection in management API');
+
+        // Force-close the connection via management API
+        $this->forceCloseConnection($connectionName);
+
+        // The server needs a moment to close the TCP connection.
+        // Retry createStream until it throws ConnectionException.
+        $maxAttempts = 10;
+        $lastException = null;
+
+        for ($attempt = 0; $attempt < $maxAttempts; $attempt++) {
+            try {
+                $connection->createStream('another-stream-' . uniqid());
+                // Still connected - wait and retry
+                usleep(200_000);
+            } catch (ConnectionException $e) {
+                $lastException = $e;
+                break;
+            }
+        }
+
+        $this->assertNotNull($lastException, 'Expected ConnectionException was not thrown after server-initiated close');
+
+        // After the operation fails, connection should be marked as disconnected
+        $this->assertFalse($connection->isConnected());
+    }
+
+    private function getStreamConnectionName(): ?string
+    {
+        $maxWait = 10;
+        $start = time();
+
+        while (time() - $start < $maxWait) {
+            $data = $this->curlGet(
+                sprintf('http://%s:%d/api/connections', self::$host, self::$managementPort)
+            );
+
+            if ($data === null) {
+                sleep(1);
+                continue;
+            }
+
+            $connections = json_decode($data, true);
+            if (!is_array($connections)) {
+                sleep(1);
+                continue;
+            }
+
+            foreach ($connections as $conn) {
+                if (!is_array($conn)) {
+                    continue;
+                }
+                if (!isset($conn['port'])) {
+                    continue;
+                }
+                if ($conn['port'] !== self::$port) {
+                    continue;
+                }
+                if (!isset($conn['protocol'])) {
+                    continue;
+                }
+                if ($conn['protocol'] !== 'stream') {
+                    continue;
+                }
+                if (!isset($conn['name'])) {
+                    continue;
+                }
+                if (!is_string($conn['name'])) {
+                    continue;
+                }
+
+                return $conn['name'];
+            }
+
+            sleep(1);
+        }
+
+        return null;
+    }
+
+    private function forceCloseConnection(string $name): void
+    {
+        $url = sprintf(
+            'http://%s:%d/api/connections/%s',
+            self::$host,
+            self::$managementPort,
+            rawurlencode($name)
+        );
+
+        $this->curlDelete($url);
+    }
+
+    private function curlGet(string $url): ?string
+    {
+        $cmd = sprintf(
+            'curl -sf -u guest:guest %s 2>/dev/null',
+            escapeshellarg($url)
+        );
+
+        $output = [];
+        $returnCode = 0;
+        exec($cmd, $output, $returnCode);
+
+        if ($returnCode !== 0 || $output === []) {
+            return null;
+        }
+
+        return implode("\n", $output);
+    }
+
+    private function curlDelete(string $url): void
+    {
+        $cmd = sprintf(
+            'curl -sf -u guest:guest -X DELETE %s >/dev/null 2>/dev/null',
+            escapeshellarg($url)
+        );
+
+        exec($cmd);
+    }
+}

